### DISCUSSION TITLE
Error uploading the code

### DISCUSSION THREAD

**Discussion by michaelhsapp** [micbay/RaceTimerAndController#19] [Q&A] [answered]
Hello, I'm almost there and looking forward to using the Lap Timer. 

I received the following error when uploading to the arduino. Do you have any guidance to resolve this?

C:\Users\Admin\AppData\Local\Temp\ccYGCnfb.ltrans0.ltrans.o: In function `loop':
C:\Users\Admin\Downloads\RaceTimerAndController/RaceTimerAndController.ino:1933: undefined reference to `updatePlayRtttl()'
C:\Users\Admin\Downloads\RaceTimerAndController/RaceTimerAndController.ino:1980: undefined reference to `stopPlayRtttl()'
C:\Users\Admin\Downloads\RaceTimerAndController/RaceTimerAndController.ino:2123: undefined reference to `startPlayRtttlPGM(unsigned char, char const*, void (*)())'
C:\Users\Admin\Downloads\RaceTimerAndController/RaceTimerAndController.ino:2660: undefined reference to `stopPlayRtttl()'
C:\Users\Admin\Downloads\RaceTimerAndController/RaceTimerAndController.ino:2665: undefined reference to `startPlayRtttlPGM(unsigned char, char const*, void (*)())'
C:\Users\Admin\Downloads\RaceTimerAndController/RaceTimerAndController.ino:2700: undefined reference to `stopPlayRtttl()'
C:\Users\Admin\Downloads\RaceTimerAndController/RaceTimerAndController.ino:2705: undefined reference to `startPlayRtttlPGM(unsigned char, char const*, void (*)())'
collect2.exe: error: ld returned 1 exit status

exit status 1

Compilation error: exit status 1

**Comment by micbay** [comment]
Hello, it looks like the PlayRttl library is missing, or the one installed is not of a compatible version. Newer versions of this library changed the names of the functions and thus are incompatible with the race controller code. Make sure that you have the 'PlayRtttl' Arduino library installed, and that the version being used is v1.4.2 and not 2.x or higher.

You can double check the last known compatible versions for the various libraries used in this project in the '[Software Configuration'](https://github.com/micbay/RaceTimerAndController?tab=readme-ov-file#software-configuration) section.

You should be able to select the version to be installed when you add it to your Arduino IDE vis the Arduino Library Manager (Tools -> Manage Libraries...).

**Reply by michaelhsapp** [reply]
Thank you. I'll be back in front of the project later this evening, will install the correct library and report back.

Stay tuned

**Reply by michaelhsapp** [reply]
Good afternoon, thanks for the help with the PlayRtttl library, I installed the correct version and was successful uploading the code.

I have a few more issues to debug:
1 - The "Timed Race" ends immediately after it starts, regardless of the set duration.
2 - I'm using IR sensors to detect the laps, I hear the buzzer beep when that signal is tripped, but the laps are not being counted in the system.
3 - Not all characters in the led bars are displayed. I think there is an edit to that library and will keep reading about that.

Do you have any suggestions?

**Reply by micbay** [reply, marked as answer]
Yeah, I think your character issue is that you haven't updated the character map, which is ([discussed here](https://github.com/micbay/RaceTimerAndController#updated-chartable-to-replace-the-one-found-in-ledcontrolh)) in the section about the racer, 7-segment displays. That will add just a couple letters like 'S' or 'Z' and those that look like numbers. However, there are a few letters that cannot be displayed in any possible form using 7-segment displays. Someday I may update this project to handle 14 segment displays and then the full alphabet will be available.

Regarding the 'Timed Race' issue you are seeing, I don't have a guess without more information, I have not seen that problem. If you can post any screen shots or video of the problem occurring on the display, I might be able to note something. Did you make any changes to the default configuration via the 'localSettings.h', perhaps a corner case has been hit, or this is a symptom of a memory overrun. However, if it's a memory issue, I'd expect it to affect standard to X laps racing too.

To test if memory is an issue, try reducing the number of racer names used and the number of stored laps from the default of 10 to 5.

In your 'localSettings.h' (which is a copy and rename of the included 'example.localSettings.h', see the [settings section](https://github.com/micbay/RaceTimerAndController#customizing-ui-text-and-general-controller-settings))
uncomment and edit these settings.
```
#define DEFAULT_MAX_STORED_LAPS 5

#define RACER_NAMES_LIST {TEXT_OFF, "Lucien", "Zoe", "Elise", "John"}
#define RACER_SONGS_LIST {disabledTone, starWarsImperialMarch, takeOnMeMB, airWolfTheme, tmnt1}
```
When you upload the sketch it should let you know how much memory is being used. There needs to be some room left for RAM during operation so if it's too close to 100% this could be the issue.

Regarding the sensors, same thing as with the 'Timed Race' issue, hard to guess as what might be going on. Perhaps try using buttons to trigger laps until everything is working, just to rule out the sensor as a factor. Note that the first trigger starts the timing of the 1st lap, so at least 2 triggers are needed before the lapcount starts iterating. Perhaps the controller is detecting the trigger, but for some reason is not able to determine which pin it came from so does not apply it to one of the racer's lapcounts. This could point to possible wiring or IDE board settings as being an issue.

**Some General Arduino Steps to troubleshoot**
Verify your board selection in the Arduino IDE (Tools->Board), and the bootloader selected (Tools->Processor) matches what is on your board. You might even just reflash the bootloader on the Arduino to make sure it is not corrupted. See [these Arduino docs](https://docs.arduino.cc/built-in-examples/arduino-isp/ArduinoISP/) for some guidance on that. Some boards may work partially with the wrong compiler settings, I had some issues that reflashing the bootloader on my Nanos solved.
Double check your pinouts, perhaps something is not wired to the right pin, an unintended reference ground mismatch, or ground loop causing aberrant behavior. Make all the power source grounds are wired together so they are at the same potential.

**Reply by michaelhsapp** [reply]
I really like this lap counter / timer, it's great!

The issue mentioned above has been solved. The 'Start' & 'Pause' buttons were wired incorrectly which resulted in them acting like they were always pressed, diagnosing it was more difficult because I hadn't corrected the character / library steps yet. But now reporting all is solved and working properly, including the IR lane detectors.

A few pictures are attached.

![Slot Car #1](https://github.com/micbay/RaceTimerAndController/assets/73416980/5a9243e8-fca5-4993-a209-63774d98f7a0)
![Slot Car #2](https://github.com/micbay/RaceTimerAndController/assets/73416980/2efe8628-51fe-486b-97d7-57efa51a2b2f)
![Slot Car #3](https://github.com/micbay/RaceTimerAndController/assets/73416980/10910e86-c914-4c7c-b53a-902519d6d617)

**Reply by micbay** [reply]
Cool! Glad you got it figured out, nice portable setup, looks great.

When you create a 'discussion', you can assign it a flag, like Q&A, Ideas, etc. as well as an option 'Show and Tell'. Please feel free to create a show and tell discussion for your project and post any pics, and build details you'd like. It'd be great to build a library of example implementations others can get ideas, and inspiration from.
